test/integration_test_wrapper.py

```diff
@@ -0,0 +1,117 @@
+
+
+TEST_EXIT_DROPIN = """\
+[Unit]
+SuccessAction=exit
+FailureAction=exit
+"""
+
+
+BUILTIN_KERNEL_COMMAND_LINE_ARGS = (
+    'systemd.hostname=H',
+    'rw',
+    'systemd.default_device_timeout_sec=20',
+    'systemd.early_core_pattern=/core',
+    'systemd.firstboot=no',
+    'noresume',
+    'oops=panic',
+    'panic=1',
+)
+
+parser = ArgumentParser(description=__doc__)
+parser.add_argument('--test-name', required=True)
+parser.add_argument('--mkosi-image-name', required=True)
+parser.add_argument('--mkosi-output-path', required=True, type=Path)
+parser.add_argument('--test-number', required=True)
+parser.add_argument('--setup-selinux', default=False, action='store_true')
+parser.add_argument('--skip-shutdown', default=False, action='store_true')
+parser.add_argument('mkosi_args', nargs="*")
+
+def main():
+    logging_basic_config(level=LOG_DEBUG)
+    args = parser.parse_args()
+
+    test_unit_name = f"testsuite-{args.test_number}.service"
+    # Machine names shouldn't have / since it's used as a file name
+    # and it must be a valid hostname so 64 chars max
+    machine_name = args.test_name.replace('/', '_')[:64]
+
+    log.debug(f"test name: {args.test_name}\n"
+              f"test number: {args.test_number}\n"
+              f"image: {args.mkosi_image_name}\n"
+              f"mkosi output path: {args.mkosi_output_path}\n"
+              f"mkosi args: {args.mkosi_args}\n"
+              f"skip shutdown: {args.skip_shutdown}")
+
+    with ExitStack() as stack:
+
+        journal_file = Path(f"{machine_name}.journal").absolute()
+        log.debug(f"Capturing journal to {journal_file}")
+
+        mkosi_args = [
+            'mkosi',
+            '--directory', Path('..').resolve(),
+            '--output-dir', args.mkosi_output_path.absolute(),
+            '--machine', machine_name,
+            '--image', args.mkosi_image_name,
+            '--format=disk',
+            '--ephemeral',
+            '--forward-journal', journal_file,
+            '--kernel-command-line-extra',
+            ' '.join([
+                *BUILTIN_KERNEL_COMMAND_LINE_ARGS,
+                *(
+                    ['apparmor=0', 'selinux=1', 'enforcing=0', 'lsm=selinux']
+                    if args.setup_selinux
+                    else ['apparmor=0', 'selinux=0', 'enforcing=0']
+                ),
+                f"SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-{args.test_number}.units:/usr/lib/systemd/tests/testdata/units:",
+                'systemd.unit=testsuite.target',
+                f"systemd.wants={test_unit_name}",
+            ]),
+            *(
+                [f"--credential=systemd.unit-dropin.{test_unit_name}={quote(TEST_EXIT_DROPIN)}"]
+                if not args.skip_shutdown
+                else []
+            ),
+            *args.mkosi_args,
+        ]
+
+        mkosi_args += ['qemu']
+
+        if log.isEnabledFor(LOG_DEBUG):
+            log.debug(f"Running {' '.join(quote(str(a)) for a in mkosi_args)}")
```

Comment:
I think this can just be `logging.debug()` without the extra check, the statement will do nothing if the log level isn't set to `DEBUG`

Reply:
This was to not perform the string quoting unless logging is enabled since it's the only case where something non-trivial was done in the logging that you'd pay for even without logging enabled (though it's not currently possible to turn it off).

I'll take the recommendation that a bit of string handling isn't heavy enough computation to worry about.